### PR TITLE
Spectral Shaping: Depthwise Conv Filter on GatedMLP Activations

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The GatedMLP modules in each TransolverBlock apply: x_out = gate(x) * value(x). After this nonlinearity, features can develop high-frequency "spectral junk" in the feature dimension that competes with the physics signal. "Spectral Shaping" (Fanaskov & Oseledets, ICLR 2025) showed that a learned depthwise 1D conv after nonlinearities — acting as a trainable low-pass/bandpass filter — significantly improves surrogate stability on fluid simulation tasks (40-60% spectral leakage reduction on Navier-Stokes).

**Proposed:** Add a depthwise 1D conv (kernel_size=3) in the feature dimension after each GatedMLP gate operation, initialized to a Gaussian blur kernel. This filters spectral artifacts before the slice-scatter step — the most information-dense bottleneck in the architecture.

Very lightweight: 576 extra parameters per TransolverBlock (1,728 total for 3 blocks).

## Instructions

**Step 1 — Add SpectralShaping module:**
```python
class SpectralShaping(nn.Module):
    """Depthwise 1D conv in feature dimension, initialized to Gaussian blur."""
    def __init__(self, n_features, kernel_size=3):
        super().__init__()
        self.conv = nn.Conv1d(n_features, n_features, kernel_size,
                              padding=kernel_size // 2, groups=n_features, bias=False)
        # Initialize to Gaussian blur: [0.25, 0.5, 0.25] for k=3
        with torch.no_grad():
            if kernel_size == 3:
                self.conv.weight.data.fill_(0)
                self.conv.weight.data[:, 0, 0] = 0.25
                self.conv.weight.data[:, 0, 1] = 0.50
                self.conv.weight.data[:, 0, 2] = 0.25

    def forward(self, x):
        # x: [B*N, D] or [B, N, D] — apply conv along feature dim
        if x.dim() == 2:
            return self.conv(x.unsqueeze(-1)).squeeze(-1)
        else:  # [B, N, D]
            B, N, D = x.shape
            x = x.reshape(B * N, D).unsqueeze(-1)  # [BN, D, 1]
            x = self.conv(x).squeeze(-1)  # [BN, D]
            return x.reshape(B, N, D)
```

**Wait — correction.** The conv should operate along the feature dimension. Since features are in the last dim, we need to transpose:
```python
def forward(self, x):
    # x: [B, N, D]
    # Conv1d expects [batch, channels, length]
    # We want to filter along D (feature dim) treating each spatial point independently
    B, N, D = x.shape
    x = x.reshape(B * N, 1, D)    # [BN, 1, D]
    # Use a regular Conv1d with 1 in-channel, 1 out-channel
    x = self.conv(x)               # [BN, 1, D]
    return x.reshape(B, N, D)
```

Actually, simplest approach — use a 1D Conv with groups=1, in_channels=1, out_channels=1:
```python
class SpectralShaping(nn.Module):
    def __init__(self, kernel_size=3):
        super().__init__()
        self.conv = nn.Conv1d(1, 1, kernel_size, padding=kernel_size // 2, bias=False)
        with torch.no_grad():
            if kernel_size == 3:
                self.conv.weight.data[0, 0] = torch.tensor([0.25, 0.50, 0.25])

    def forward(self, x):
        shape = x.shape
        x = x.reshape(-1, 1, shape[-1])  # [B*N, 1, D]
        x = self.conv(x)                  # [B*N, 1, D]
        return x.reshape(shape)
```

**Step 2 — Add CLI flags:**
```python
parser.add_argument('--spectral_shaping', action='store_true',
                    help='Add spectral shaping conv after GatedMLP')
parser.add_argument('--spectral_shaping_kernel', type=int, default=3,
                    help='Kernel size for spectral shaping conv')
```

**Step 3 — Integrate into GatedMLP or TransolverBlock.** After the `gate(x) * value(x)` operation in each GatedMLP, apply spectral shaping:
```python
if args.spectral_shaping:
    self.spec_shape = SpectralShaping(kernel_size=args.spectral_shaping_kernel)

# In forward:
x = gate * value
if hasattr(self, 'spec_shape'):
    x = self.spec_shape(x)
```

**Step 4 — Run 2 seeds** using `--wandb_group frieren-spectral-shaping`:

```bash
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/spec-shape-k3" --wandb_group frieren-spectral-shaping \
  --spectral_shaping --spectral_shaping_kernel 3 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(repeat with `--seed 73`)

Report p_in, p_oodc, p_tan, p_re (2-seed avg).

## Baseline

| Metric | Baseline (2-seed avg) |
|--------|----------------------|
| p_in | 13.05 |
| p_oodc | 7.70 |
| **p_tan** | **28.60** |
| p_re | 6.55 |

Baseline W&B runs: d7l91p0x (seed 42), j9btfx09 (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Summary

Spectral shaping produced **mixed results**. Seed 42 matched baseline p_tan exactly (28.6), but seed 73 regressed significantly (29.9), giving a 2-seed avg of 29.25 (+2.3% vs baseline). The high seed variance suggests the technique is not robustly beneficial.

### Metrics

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B |
|--------|------|------|--------|-------|------|-----|
| spec-shape k=3 | 42 | 13.4 | 7.8 | **28.6** | 6.6 | `i469q07m` |
| spec-shape k=3 | 73 | 13.6 | 7.8 | 29.9 | 6.6 | `4ta4863c` |
| **avg** | — | **13.5** | **7.8** | **29.25** | **6.6** | |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 |

| Metric | Δ vs baseline |
|--------|---------------|
| p_in | +3.4% (worse) |
| p_oodc | +1.3% (noise) |
| p_tan | **+2.3% (worse)** |
| p_re | +0.8% (noise) |

### Implementation

Added `SpectralShaping` module — a learnable 3-tap filter `[w0, w1, w2]` initialized to Gaussian blur `[0.25, 0.50, 0.25]`, applied along the feature dimension after the MLP in each TransolverBlock. Implemented as shifted-tensor weighted sums for torch.compile efficiency (initial Conv1d implementation caused 2.5x slowdown; the tensor-slicing version runs at baseline speed: 73-74s/epoch vs 70-72s baseline).

### VRAM & Speed
- 46.2-47.2 GB (vs 44-46 GB baseline, +1-2 GB)
- 73-74s/epoch post-compilation (≈baseline)
- 146-147 epochs in 180 min

### Commands

```bash
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/spec-shape-k3-s42-v2" --wandb_group frieren-spectral-shaping \
  --spectral_shaping --spectral_shaping_kernel 3 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### What happened

The spectral shaping filter is extremely lightweight (3 learnable params per block = 9 total), and the Gaussian-blur initialization means it starts as a mild smoothing operation. Seed 42 reached baseline-level p_tan (28.6, exactly matching baseline), suggesting the feature-smoothing concept has merit for some initialization trajectories. However, seed 73 regressed to 29.9, indicating the filter disrupts feature representations for other random seeds.

The high variance between seeds (28.6 vs 29.9 = 4.5% spread, much wider than baseline's 28.3-28.9 = 2.1% spread) suggests the learned filter interacts unpredictably with the random initialization of surrounding layers. The filter weights are free to evolve away from the initial Gaussian blur, and for seed 73 they may have converged to a less beneficial configuration.

### Suggested follow-ups

1. **Fix the filter weights (non-learnable):** Keep `[0.25, 0.50, 0.25]` frozen (no gradient). This removes the optimization instability while retaining the smoothing benefit. If seed 42's result holds with a frozen filter, this would be a zero-cost improvement.
2. **Apply only to the last block:** The tandem transfer metric is most sensitive to the last block's representations. Spectral shaping on just the last block would be a more targeted intervention.
3. **Try kernel_size=5:** A wider filter `[0.0625, 0.25, 0.375, 0.25, 0.0625]` (Gaussian k=5) might provide more robust smoothing.
4. **Residual spectral shaping:** `x + alpha * spec_shape(x)` with small alpha (0.1) — gentler perturbation that preserves the original features more.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2176 because ADVISOR: Closing — spectral shaping is unstable across seeds.

**s42:** p_tan=28.59 (matches baseline exactly)
**s73:** p_tan=29.95 (+4.7%)
**avg:** p_tan=29.27 (+2.3%)

Seed variance of 4.5% (28.59 vs 29.95) is double the baseline spread (2.1%). The feature smoothing concept isn't robust — the learned filter evolves unpredictably depending on initialization.

Your frozen-filter follow-up is interesting but even the best seed (28.59) only matches baseline, it doesn't beat it. The smoothing operation is either neutral or harmful — there's no headroom here.

The torch.compile-compatible tensor slicing implementation is good engineering work. The architectural insight (feature-space smoothing) may be revisited if combined with other changes, but as a standalone experiment it doesn't have enough signal..
